File: csharp/src/SpringNBTLibrary/SpringNbtException.cs
namespace SpringNBTLibrary;

/// <summary>
/// 本ライブラリが送出する例外。分類は <see cref="Code"/> で判別する。
/// </summary>
/// <remarks>仕様: <c>docs/spec/00-conventions.md</c> 4章 / <c>docs/adr/0005-unified-error-model.md</c></remarks>
public sealed class SpringNbtException : Exception
{
    /// <summary>分類とメッセージから例外を作る。</summary>
    public SpringNbtException(ErrorCode code, string message)
        : base(message)
    {
        Code = code;
    }

    /// <summary>原因となった下位の例外を保持して例外を作る。</summary>
    public SpringNbtException(ErrorCode code, string message, Exception innerException)
        : base(message, innerException)
    {
        Code = code;
    }

    /// <summary>エラーの分類。</summary>
    public ErrorCode Code { get; }

    /// <inheritdoc/>
    public override string ToString()
    {
        return $"[{Code.AsString()}] {Message}";
    }

    /// <summary><see cref="ErrorCode.MalformedData"/> の例外を作る。</summary>
    public static SpringNbtException Malformed(string message)
    {
        return new SpringNbtException(ErrorCode.MalformedData, message);
    }

    /// <summary><see cref="ErrorCode.InvalidArgument"/> の例外を作る。</summary>
    public static SpringNbtException InvalidArgument(string message)
    {
        return new SpringNbtException(ErrorCode.InvalidArgument, message);
    }

    /// <summary><see cref="ErrorCode.UnexpectedTagType"/> の例外を作る。</summary>
    public static SpringNbtException UnexpectedTagType(string message)
    {
        return new SpringNbtException(ErrorCode.UnexpectedTagType, message);
    }

    /// <summary><see cref="ErrorCode.LimitExceeded"/> の例外を作る。</summary>
    public static SpringNbtException LimitExceeded(string message)
    {
        return new SpringNbtException(ErrorCode.LimitExceeded, message);
    }

    /// <summary><see cref="ErrorCode.UnsupportedFeature"/> の例外を作る。</summary>
    public static SpringNbtException UnsupportedFeature(string message)
    {
        return new SpringNbtException(ErrorCode.UnsupportedFeature, message);
    }
}

File: java/src/main/java/io/github/scriptarts/springnbt/SpringNbtException.java
package io.github.scriptarts.springnbt;

/**
 * 本ライブラリが送出する例外。分類は {@link #code()} で判別する。
 *
 * <p>検査例外にしていないのは、4言語でメソッドのシグネチャを揃えるため。
 * 入出力エラーも {@link ErrorCode#IO} でラップして送出する。
 *
 * <p>仕様: {@code docs/spec/00-conventions.md} 4章 / {@code docs/adr/0005-unified-error-model.md}
 */
public final class SpringNbtException extends RuntimeException {

    private static final long serialVersionUID = 1L;

    private final ErrorCode code;

    /**
     * 分類とメッセージから例外を作る。
     *
     * @param code    エラーの分類
     * @param message 説明
     */
    public SpringNbtException(ErrorCode code, String message) {
        super(message);
        this.code = code;
    }

    /**
     * 原因となった下位の例外を保持して例外を作る。
     *
     * @param code    エラーの分類
     * @param message 説明
     * @param cause   原因
     */
    public SpringNbtException(ErrorCode code, String message, Throwable cause) {
        super(message, cause);
        this.code = code;
    }

    /**
     * エラーの分類。
     *
     * @return 分類
     */
    public ErrorCode code() {
        return code;
    }

    @Override
    public String toString() {
        return "[" + code.asString() + "] " + getMessage();
    }

    /**
     * {@link ErrorCode#MALFORMED_DATA} の例外を作る。
     *
     * @param message 説明
     * @return 例外
     */
    public static SpringNbtException malformed(String message) {
        return new SpringNbtException(ErrorCode.MALFORMED_DATA, message);
    }

    /**
     * {@link ErrorCode#INVALID_ARGUMENT} の例外を作る。
     *
     * @param message 説明
     * @return 例外
     */
    public static SpringNbtException invalidArgument(String message) {
        return new SpringNbtException(ErrorCode.INVALID_ARGUMENT, message);
    }

    /**
     * {@link ErrorCode#UNEXPECTED_TAG_TYPE} の例外を作る。
     *
     * @param message 説明
     * @return 例外
     */
    public static SpringNbtException unexpectedTagType(String message) {
        return new SpringNbtException(ErrorCode.UNEXPECTED_TAG_TYPE, message);
    }

    /**
     * {@link ErrorCode#LIMIT_EXCEEDED} の例外を作る。
     *
     * @param message 説明
     * @return 例外
     */
    public static SpringNbtException limitExceeded(String message) {
        return new SpringNbtException(ErrorCode.LIMIT_EXCEEDED, message);
    }

    /**
     * {@link ErrorCode#UNSUPPORTED_FEATURE} の例外を作る。
     *
     * @param message 説明
     * @return 例外
     */
    public static SpringNbtException unsupportedFeature(String message) {
        return new SpringNbtException(ErrorCode.UNSUPPORTED_FEATURE, message);
    }
}
